stratum/hal/lib/tai/taiadapterhost.cc

```diff
@@ -0,0 +1,252 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+#include "stratum/hal/lib/tai/taiadapterhost.h"
+
+#include <string>
+#include <vector>
+#include <utility>
+
+#include "stratum/glue/logging.h"
+#include "stratum/lib/utils.h"
+
+#include "absl/memory/memory.h"
+
+#include "stratum/hal/lib/tai/hostinterface.h"
+#include "stratum/hal/lib/tai/module.h"
+#include "stratum/hal/lib/tai/networkinterface.h"
+
+namespace stratum {
+namespace hal {
+namespace tai {
+
+/*!
+ * \brief modules_location variable used to store returned from TAI module
+ * location
+ */
+static std::vector<std::string> modules_location;
+
+/*!
+ * \brief module_presence function will be called once for each module present
+ * toward the end of the tai_api_initialze function, and then whenever there is
+ * a change
+ */
+static void module_presence(bool /*present*/, char* location) {
+  // \param present should be used to delete the module that was unplugged (I
+  // think)
+  modules_location.emplace_back(location);
+}
+
+TAIAdapterHost::TAIAdapterHost()
+    : path_rule_({{tai_object_type_t::TAI_OBJECT_TYPE_MODULE},
+                  {tai_object_type_t::TAI_OBJECT_TYPE_MODULE,
+                   tai_object_type_t::TAI_OBJECT_TYPE_NETWORKIF},
+                  {tai_object_type_t::TAI_OBJECT_TYPE_MODULE,
+                   tai_object_type_t::TAI_OBJECT_TYPE_HOSTIF}}) {
+  LOG(INFO) << "Initialize TAIAdapterHost";
+  tai_service_method_table_t services;
+  services.module_presence = module_presence;
+
+  auto status = tai_api_initialize(0, &services);
+  if (TAI_STATUS_SUCCESS != status) {
+    LOG(ERROR) << "Failed to initialize TAIAdapterHost. Error status: "
+               << status;
+    return;
+  }
+
+  status =
+      tai_api_query(TAI_API_MODULE, reinterpret_cast<void**>(&api_.module_api));
+  if (TAI_STATUS_SUCCESS != status) {
+    LOG(ERROR) << "Failed to query MODULE API" << std::endl;
+    return;
+  }
+
+  status = tai_api_query(TAI_API_NETWORKIF,
+                         reinterpret_cast<void**>(&api_.netif_api));
+  if (TAI_STATUS_SUCCESS != status) {
+    LOG(ERROR) << "Failed to query NETWORKIF API" << std::endl;
+    return;
+  }
+
+  status =
+      tai_api_query(TAI_API_HOSTIF, reinterpret_cast<void**>(&api_.hostif_api));
+  if (TAI_STATUS_SUCCESS != status) {
+    LOG(ERROR) << "Failed to query HOSTIF API" << std::endl;
+    return;
+  }
+
+  for (const auto& location : modules_location) {
```

Comment:
You can't be sure that the presence callback as been called here. This must happen later. I'd make that a separate function.